app/services/display_ads/sample_relevant_ads.rb

```diff
@@ -0,0 +1,77 @@
+module DisplayAds
+  class SampleRelevantAds
+    attr_reader :filtered_display_ads
+
+    def self.call(...)
+      new(...).call
+    end
+
+    def initialize(display_ads:, area:, user_signed_in:, article_tags:)
+      @filtered_display_ads = display_ads
+      @area = area
+      @user_signed_in = user_signed_in
+      @article_tags = article_tags
+    end
+
+    def call
+      @filtered_display_ads = approved_and_published_ads
+
+      if @article_tags.any?
+        @filtered_display_ads = tagged_post_comment_ads
+      end
+
+      if @article_tags.blank?
+        @filtered_display_ads = untagged_post_comment_ads
+      end
+
+      @filtered_display_ads = authenticated_ads
+      @filtered_display_ads.order(success_rate: :desc)
+
+      @filtered_display_ads = sample_ads
+    end
+
+    private
+
+    def approved_and_published_ads
+      @filtered_display_ads.approved_and_published.where(placement_area: @area)
+    end
+
+    def tagged_post_comment_ads
+      display_ads_with_targeted_article_tags = @filtered_display_ads.cached_tagged_with_any(@article_tags)
+
+      untagged_post_comment_ads.or(display_ads_with_targeted_article_tags)
+    end
+
+    def untagged_post_comment_ads
+      @filtered_display_ads.where(cached_tag_list: "")
+    end
+
+    def authenticated_ads
+      if @user_signed_in
+        @filtered_display_ads.where(display_to: %w[all logged_in])
+      else
+        @filtered_display_ads.where(display_to: %w[all logged_out])
+      end
+    end
+
+    # Business Logic Context:
```

Comment:
This may be an anti-pattern but I added this pretty long comment here because I struggled to rationalize about this piece of code for a while. Hence, I thought it may be useful to drop in here. If we thing its a terrible idea, I'm happy to remove it.

Reply:
I think it's useful :)